fs/error_tools.py

```diff
@@ -84,7 +84,8 @@ def __exit__(
             _errno = exc_value.errno
             fserror = os_errors.get(_errno, errors.OperationFailed)
             if _errno == errno.EACCES and sys.platform == "win32":
-                if getattr(exc_value, "args", None) == 32:  # pragma: no cover
+                error_args = getattr(exc_value, "args", (None,))
+                if error_args and error_args[0] == 32:  # pragma: no cover
```

Comment:
A non-empty tuple (even `(None,)`) evaluates to `True`, so is the `error_args and` part needed in the `if`?
```python
>>> bool((None,))
True
```

Reply:
I included this because `OSError` can be thrown without arguments, in which case `args` will be empty.

Reply:
I wonder if it's worth just returning a plain None as the `getattr` default case then? i.e. `error_args = getattr(exc_value, "args", None)`

Reply:
That was my first thought, but then I figured if we add MyPy there's a good chance it would get angry about that, so I used a tuple instead.

Reply:
```suggestion
                error_args = getattr(exc_value, "args", (None,))
                if error_args and error_args[0] == 13:  # pragma: no cover
```
should check if that argument is 13 instead

Reply:
Also this line https://github.com/PyFilesystem/pyfilesystem2/pull/491/files#R63 
`FILE_ERRORS[13] = errors.FileExpected `
I don't get it seems wrong since the error code 13 should be resource locked error and I don't think it would ever throw file expected in this situation anyway since fserror gets overwritten later. Actually you could probably just do:

`FILE_ERRORS[13] = errors.ResourceLocked`

and not do all that if _errno == errno.EACCES and checking stuff there instead. I think this is right?

Reply:
@dargueta yeah this works too it seems:
```
  if _WINDOWS_PLATFORM:  # pragma: no cover
      DIR_ERRORS[13] = errors.DirectoryExpected
      DIR_ERRORS[267] = errors.DirectoryExpected
      FILE_ERRORS[13] = errors.ResourceLocked

  def __init__(self, opname, path, directory=False):
      # type: (Text, Text, bool) -> None
      self._opname = opname
      self._path = path
      self._directory = directory
      
  def __enter__(self):
      # type: () -> _ConvertOSErrors
      return self

  def __exit__(
      self,
      exc_type,  # type: Optional[Type[BaseException]]
      exc_value,  # type: Optional[BaseException]
      traceback,  # type: Optional[TracebackType]
  ):
      # type: (...) -> None
      os_errors = self.DIR_ERRORS if self._directory else self.FILE_ERRORS
      if exc_type and isinstance(exc_value, EnvironmentError):
          _errno = exc_value.errno
          fserror = os_errors.get(_errno, errors.OperationFailed)
          reraise(fserror, fserror(self._path, exc=exc_value), traceback)
```

just might wanna look as to why this line was added:
`FILE_ERRORS[13] = errors.FileExpected`

Reply:
So here's the funny thing -- according to [this](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-)

Error 32 on Windows is `ERROR_SHARING_VIOLATION`: "The process cannot access the file because it is being used by another process." This makes the most sense for the original code of the `if` statement here.

Error 13 is `ERROR_INVALID_DATA` on Windows which might be the correct one for "file expected" but I'm not sure, given there are things like `ERROR_DIRECTORY_NOT_SUPPORTED`.

_However_

POSIX error code 13 is `EACCES`, which is what you'd expect from a failed access due to a file being locked. So it looks like we should be checking `winerror` against 32, not errno?

```py
windows_error = getattr(exc_value, "winerror", 0)
exception_args = getattr(exc_value, "args", None) or (0,)
if windows_error == 32 or exception_args[0] == errno.EACCES:
    # proper exception type is errors.ResourceLocked
```

Reply:
@dargueta sorry for not getting back sooner I got really busy with some stuff and forgot. Idk how much more you've looked at this since then but I found some stuff that helped me understand it better. 

I was playing around with another test that raised this:
`pywintypes.error: (32, 'CreateFileW', 'The process cannot access the file because it is being used by another process.')
`
which is the windows error you'd expect. 

Seems to me that there is a step in between where you are expected to convert this windows error to oserror yourself by doing something like:
`exception = OSError(32, "resource locked")
exception.errno = 13`
since an oserror like that won't get thrown but does seem jank since you would expect args[0] to match errno at least I think its supposed to?
 
If this is the expected use here and I think it is, would be nice to include a winerror to oserror decorator or something like it in error tools or a comment explaining it or improving this to convert the winerror directly to fserror. Or at least that's my opinion on it.

Reply:
I _think_ the `windows_error = getattr(exc_value, "winerror", 0)` bit handles it. I forgot to update the unit tests with the latest change though. 😅